File: gui/backend/routers/hypothesis.py
"""Hypothesis Tests router."""

import sys
import math
from pathlib import Path
from typing import Any, Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "src"))

from reliability.Hypothesis_tests import (
    one_sample_t,
    two_sample_t,
    paired_t,
    mann_whitney,
    wilcoxon_signed_rank,
    chi_square_gof,
    chi_square_independence,
    binomial_test,
    one_way_anova,
    kruskal_wallis,
    friedman,
    anova_factorial,
    repeated_measures_anova,
    mixed_anova,
)

router = APIRouter()


# ---------------------------------------------------------------------------
# Float sanitizer — converts nan/inf to None recursively
# ---------------------------------------------------------------------------

from utils import safe as _safe


# ---------------------------------------------------------------------------
# Inline Pydantic schemas
# ---------------------------------------------------------------------------

class RunRequest(BaseModel):
    test: str
    # One-sample
    data: Optional[list[float]] = None
    popmean: Optional[float] = 0.0
    # Two-sample / paired / nonparametric
    group_a: Optional[list[float]] = None
    group_b: Optional[list[float]] = None
    equal_var: Optional[bool] = False
    # k-group
    groups: Optional[list[list[float]]] = None
    # Chi-square GOF
    observed: Optional[list[float]] = None
    expected: Optional[list[float]] = None
    # Chi-square independence
    table: Optional[list[list[float]]] = None
    # Binomial
    successes: Optional[int] = None
    n: Optional[int] = None
    p: Optional[float] = 0.5
    # Common
    alpha: float = 0.05
    alternative: str = "two-sided"


class AnovaRequest(BaseModel):
    response: list[float]
    factors: dict[str, list[str]]
    factor_names: list[str]
    alpha: float = 0.05


class RMAnovaRequest(BaseModel):
    data: list[list[float]]   # shape: subjects x conditions
    alpha: float = 0.05


class MixedAnovaRequest(BaseModel):
    values: list[float]
    subjects: list[str]
    between_factor: list[str]
    within_factor: list[str]
    alpha: float = 0.05


# ---------------------------------------------------------------------------
# Dispatch map for /run
# ---------------------------------------------------------------------------

TEST_DISPATCH = {
    # Parametric
    "one_sample_t": lambda req: one_sample_t(
        req.data, popmean=req.popmean or 0.0, alpha=req.alpha, alternative=req.alternative
    ),
    "two_sample_t": lambda req: two_sample_t(
        req.group_a, req.group_b, alpha=req.alpha,
        equal_var=req.equal_var or False, alternative=req.alternative
    ),
    "paired_t": lambda req: paired_t(
        req.group_a, req.group_b, alpha=req.alpha, alternative=req.alternative
    ),
    # Non-parametric
    "mann_whitney": lambda req: mann_whitney(
        req.group_a, req.group_b, alpha=req.alpha, alternative=req.alternative
    ),
    "wilcoxon_signed_rank": lambda req: wilcoxon_signed_rank(
        req.group_a, req.group_b, alpha=req.alpha, alternative=req.alternative
    ),
    # Chi-square / proportion
    "chi_square_gof": lambda req: chi_square_gof(
        req.observed, expected=req.expected or None, alpha=req.alpha
    ),
    "chi_square_independence": lambda req: chi_square_independence(
        req.table, alpha=req.alpha
    ),
    "binomial_test": lambda req: binomial_test(
        req.successes, req.n, p=req.p or 0.5,
        alpha=req.alpha, alternative=req.alternative
    ),
    # k-group
    "one_way_anova": lambda req: one_way_anova(
        req.groups, alpha=req.alpha
    ),
    "kruskal_wallis": lambda req: kruskal_wallis(
        req.groups, alpha=req.alpha
    ),
    "friedman": lambda req: friedman(
        req.groups, alpha=req.alpha
    ),
}


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------

@router.post("/run")
def run_test(req: RunRequest):
    """
    Dispatch to the appropriate hypothesis test.

    The `test` field selects which test to run. Required payload fields
    vary by test — see the module docstrings.
    """
    handler = TEST_DISPATCH.get(req.test)
    if handler is None:
        raise HTTPException(
            status_code=400,
            detail=f"Unknown test '{req.test}'. "
                   f"Valid tests: {', '.join(TEST_DISPATCH.keys())}",
        )
    try:
        result = handler(req)
        return _safe(result)
    except (ValueError, TypeError) as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Computation error: {e}")


@router.post("/anova")
def run_anova(req: AnovaRequest):
    """
    Factorial ANOVA (1-, 2-, or 3-way with interactions).

    Provide `response`, `factors` (dict of factor_name → list of level strings),
    and `factor_names` (ordered list of 1–3 factor names to include).
    """
    try:
        result = anova_factorial(
            req.response,
            factors=req.factors,
            factor_names=req.factor_names,
            alpha=req.alpha,
        )
        return _safe(result)
    except (ValueError, TypeError) as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Computation error: {e}")


@router.post("/rm-anova")
def run_rm_anova(req: RMAnovaRequest):
    """
    Repeated-measures ANOVA (one within-subject factor).

    Provide `data` as a 2D list with shape (n_subjects, n_conditions).
    """
    try:
        result = repeated_measures_anova(req.data, alpha=req.alpha)
        return _safe(result)
    except (ValueError, TypeError) as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Computation error: {e}")


@router.post("/mixed-anova")
def run_mixed_anova(req: MixedAnovaRequest):
    """
    Mixed ANOVA (one between + one within factor).

    Provide data in long format: `values`, `subjects`, `between_factor`,
    and `within_factor` (all same length).
    """
    try:
        result = mixed_anova(
            req.values,
            req.subjects,
            req.between_factor,
            req.within_factor,
            alpha=req.alpha,
        )
        return _safe(result)
    except (ValueError, TypeError) as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Computation error: {e}")

File: gui/frontend/src/api/hypothesis.ts
import { api } from './client'

// ---------------------------------------------------------------------------
// Common result shape shared by all tests
// ---------------------------------------------------------------------------

export interface HypothesisResult {
  test: string
  statistic: number | null
  p_value: number | null
  df: number | { [key: string]: number } | null
  effect_size: number | null
  effect_size_name?: string
  effect_size_direction?: string
  alpha: number
  reject_null: boolean
  alternative?: string
  interpretation: string
  // Confidence interval on the mean / mean-difference (t-tests)
  ci_lower?: number | null
  ci_upper?: number | null
  ci_level?: number
  ci_on?: string
  // One-sample extras
  sample_mean?: number | null
  sample_sd?: number | null
  n?: number
  popmean?: number
  // Two-sample extras
  equal_var?: boolean
  mean_a?: number | null
  mean_b?: number | null
  sd_a?: number | null
  sd_b?: number | null
  n_a?: number
  n_b?: number
  // Paired extras
  mean_diff?: number | null
  sd_diff?: number | null
  // Binomial extras
  successes?: number
  p_null?: number
  p_observed?: number | null
  // k-group extras
  k?: number
  n_total?: number
  group_means?: (number | null)[]
  group_sds?: (number | null)[]
  group_ns?: number[]
  group_medians?: (number | null)[]
  pairwise_bonferroni?: PairwiseComparison[]
  // SS decomposition (one-way ANOVA)
  ss_between?: number | null
  ss_within?: number | null
  ss_total?: number | null
  ms_between?: number | null
  ms_within?: number | null
  // ANOVA table (factorial / RM / mixed)
  anova_table?: AnovaTableRow[]
  // Mixed ANOVA sub-effects
  between_factor?: SubEffect
  within_factor?: SubEffect
  interaction?: SubEffect
  // Misc
  n_subjects?: number
  n_conditions?: number
  n_between_levels?: number
  n_within_levels?: number
  balance_note?: string
  inference_basis?: string
  p_value_uncorrected?: number | null
  sphericity?: {
    status: string
    W: number | null
    chi_square: number | null
    df: number
    p_value: number | null
    reject_sphericity: boolean | null
    epsilon_greenhouse_geisser: number | null
    epsilon_huynh_feldt: number | null
    epsilon_lower_bound: number
  }
  corrections?: Record<string, {
    epsilon: number | null
    df_conditions: number | null
    df_error: number | null
    p_value: number | null
  }>
  model?: {
    estimation: string
    covariance_structure: string
    residual_subject_df: number
    group_sizes: Record<string, number>
    repeated_covariance: number[][]
    raw_covariance_condition_number: number | null
    warnings: string[]
    assumptions: string[]
  }
  shape?: number[]
}

export interface PairwiseComparison {
  group_i: number
  group_j: number
  mean_diff: number | null
  p_value_raw: number | null
  p_value_bonferroni: number | null
  significant: boolean
}

export interface AnovaTableRow {
  source: string
  SS: number | null
  df: number | null
  MS: number | null
  F: number | null
  p_value: number | null
  partial_eta_sq?: number | null
  significant?: boolean | null
  df_den?: number | null
  method?: string
  correction?: string
  df_uncorrected?: number
}

export interface SubEffect {
  F: number | null
  p_value: number | null
  reject_null: boolean
  df_num?: number
  df_den?: number
  wald_chi_square?: number
  wald_chi_square_p?: number
  inference_method?: string
}

// ---------------------------------------------------------------------------
// /run — dispatch to any simple test
// ---------------------------------------------------------------------------

export interface RunRequest {
  test: string
  // One-sample
  data?: number[]
  popmean?: number
  // Two-sample / paired / nonparametric
  group_a?: number[]
  group_b?: number[]
  equal_var?: boolean
  // k-group
  groups?: number[][]
  // Chi-square GOF
  observed?: number[]
  expected?: number[] | null
  // Chi-square independence
  table?: number[][]
  // Binomial
  successes?: number
  n?: number
  p?: number
  // Common
  alpha?: number
  alternative?: string
}

export const runHypothesisTest = (req: RunRequest) =>
  api.post<HypothesisResult>('/hypothesis/run', req).then(r => r.data)

// ---------------------------------------------------------------------------
// /anova — factorial 1–3 way
// ---------------------------------------------------------------------------

export interface AnovaRequest {
  response: number[]
  factors: Record<string, string[]>
  factor_names: string[]
  alpha?: number
}

export const runAnova = (req: AnovaRequest) =>
  api.post<HypothesisResult>('/hypothesis/anova', req).then(r => r.data)

// ---------------------------------------------------------------------------
// /rm-anova — repeated-measures ANOVA
// ---------------------------------------------------------------------------

export interface RMAnovaRequest {
  data: number[][]   // shape: subjects x conditions
  alpha?: number
}

export const runRMAnova = (req: RMAnovaRequest) =>
  api.post<HypothesisResult>('/hypothesis/rm-anova', req).then(r => r.data)

// ---------------------------------------------------------------------------
// /mixed-anova — one between + one within factor
// ---------------------------------------------------------------------------

export interface MixedAnovaRequest {
  values: number[]
  subjects: string[]
  between_factor: string[]
  within_factor: string[]
  alpha?: number
}

export const runMixedAnova = (req: MixedAnovaRequest) =>
  api.post<HypothesisResult>('/hypothesis/mixed-anova', req).then(r => r.data)
